subsys/net/lib/aws_fota/src/aws_fota.c

```diff
@@ -0,0 +1,517 @@
+			if (err) {
+				LOG_ERR("Error when trying to start firmware"
+				       "download: %d", err);
+				return err;
+			}
+		} else if (execution_state == AWS_JOBS_IN_PROGRESS &&
+		    fota_state == APPLY_FIRMWARE) {
+			LOG_DBG("Download completed");
+			/* clean up and report status
+			 * Maybe keep applying firmware as we haven't rebooted
+			 */
+			execution_state = AWS_JOBS_SUCCEEDED;
+			update_job_execution(client, job_id, AWS_JOBS_SUCCEEDED,
+					fota_state, doc_version_number, "");
+		} else if (execution_state == AWS_JOBS_SUCCEEDED &&
+			   fota_state == APPLY_FIRMWARE) {
+			printk("Done emmit");
+			callback(AWS_FOTA_EVT_DONE);
+		}
+		return 1;
+	} else if (!strncmp(job_id_update_rejected_topic, topic,
+			    MIN(JOB_ID_UPDATE_TOPIC_MAX_LEN, topic_len))) {
+		printk("Update was rejected\n");
+		callback(AWS_FOTA_EVT_ERROR);
+	}
+	return 0;
+
+}
+
+int aws_fota_mqtt_evt_handler(struct mqtt_client *const client,
+			      const struct mqtt_evt *evt)
+{
+	int err;
+
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			return 0;
+		}
+
+		err = aws_jobs_subscribe_notify_next(client);
+		if (err) {
+			return err;
+		}
+
+		err = update_device_shadow_version(client);
+		if (err) {
+			return err;
+		}
+
+		return 0;
+		/* This expects that the application's mqtt handler will handle
+		 * any situations where you could not connect to the MQTT
+		 * broker.
+		 */
+
+	case MQTT_EVT_DISCONNECT:
+		return 0;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *p = &evt->param.publish;
+
+		err = aws_fota_on_publish_evt(client,
+					      p->message.topic.topic.utf8,
+					      p->message.topic.topic.size,
+					      payload_buf,
+					      p->message.payload.len);
+		if (err < 1) {
+			return err;
+		}
+
+		if (p->message.topic.qos == MQTT_QOS_1_AT_LEAST_ONCE) {
+			const struct mqtt_puback_param ack = {
+				.message_id = p->message_id
+			};
+
+			/* Send acknowledgment. */
+			err = mqtt_publish_qos1_ack(c, &ack);
+			if (err) {
+				return err;
+			}
+		}
+		return 1;
+
+	} break;
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBACK error %d", evt->result);
+			return 0;
+		}
+		/* check evt->param.puback.message_id */
+		/* We expect that the client pubacks*/
+		return 0;
+
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			return 0;
+		}
+		if ((fota_state == DOWNLOAD_FIRMWARE) &&
+		   (evt->param.suback.message_id == SUBSCRIBE_JOB_ID_UPDATE)) {
+			/* Client token are not used by this library */
+			err = update_job_execution(client, job_id,
+						   AWS_JOBS_IN_PROGRESS,
+						   fota_state,
+						   doc_version_number,
+						   "");
+			if (err) {
+				printk("Error when updating "
+				       "job_execution_status : %d\n", err);
+			return err;
+			}
+		}
+
+		return 0;
+
+	default:
+		/* Handling for default case? */
+		return 0;
+	}
+	return 0;
+}
+
+static void http_fota_handler(enum fota_download_evt_id evt)
+{
+	__ASSERT_NO_MSG(c != NULL);
```

Comment:
Does this make sense? I assumed that static pointers were set to `NULL` upon declaration but this could be wrong.

Reply:
I think it's fine - technically this handler shall not be called before `aws_fota_init` so there should be no harm. And reasonably placed asserts help to easily spot if something fishy is going on

Reply:
Okei :) I'll leave it then.